desc/objectives/nae_utils.py

```diff
@@ -446,3 +450,396 @@ def make_RZ_cons_1st_order(qsc, desc_eq, fix_lambda=False, N=None):
+    x1_sin = (
+        stel.Y1s_untwisted * stel.binormal_cylindrical.transpose()
+        + stel.X1s_untwisted * stel.normal_cylindrical.transpose()
+    )
+
+    x2_0 = (
+        stel.X20_untwisted * stel.normal_cylindrical.transpose()
+        + stel.Y20_untwisted * stel.binormal_cylindrical.transpose()
+        + stel.Z20_untwisted * stel.tangent_cylindrical.transpose()
+    )
+    x2_cos = (
+        stel.X2c_untwisted * stel.normal_cylindrical.transpose()
+        + stel.Y2c_untwisted * stel.binormal_cylindrical.transpose()
+        + stel.Z2c_untwisted * stel.tangent_cylindrical.transpose()
+    )
+    x2_sin = (
+        stel.X2s_untwisted * stel.normal_cylindrical.transpose()
+        + stel.Y2s_untwisted * stel.binormal_cylindrical.transpose()
+        + stel.Z2s_untwisted * stel.tangent_cylindrical.transpose()
+    )
+
+    X1Rc = x1_cos[0, :]
+    dX1Rc = np.matmul(stel.d_d_phi, X1Rc)
+    X1Rs = x1_sin[0, :]
+    dX1Rs = np.matmul(stel.d_d_phi, X1Rs)
+    X1phic = x1_cos[1, :]
+    dX1phic = np.matmul(stel.d_d_phi, X1phic)
+    X1phis = x1_sin[1, :]
+    dX1phis = np.matmul(stel.d_d_phi, X1phis)
+    X1zc = x1_cos[2, :]
+    dX1zc = np.matmul(stel.d_d_phi, X1zc)
+    X1zs = x1_sin[2, :]
+    dX1zs = np.matmul(stel.d_d_phi, X1zs)
+    X2R0 = x2_0[0, :]
+    X2phi0 = x2_0[1, :]
+    X2z0 = x2_0[2, :]
+    X2Rs = x2_sin[0, :]
+    X2phis = x2_sin[1, :]
+    X2zs = x2_sin[2, :]
+    X2Rc = x2_cos[0, :]
+    X2phic = x2_cos[1, :]
+    X2zc = x2_cos[2, :]
+
+    coeffs = {}
+
+    R_2_0 = (
+        (X1phic**2 + X1phis**2) / 4 / R0
+        + X2R0
+        + (X1Rc * X1phic * dR0) / 2 / R0**2
+        + (X1Rs * X1phis * dR0) / 2 / R0**2
+        - X2phi0 * dR0 / R0
+        - X1phic**2 * dR0**2 / 2 / R0**3
+        - X1phis**2 * dR0**2 / 2 / R0**3
+        - X1phic * dX1Rc / 2 / R0
+        - X1phis * dX1Rs / 2 / R0
+        + X1phic * dR0 * dX1phic / 2 / R0**2
+        + X1phis * dR0 * dX1phis / 2 / R0**2
+        + X1phic**2 * ddR0 / 4 / R0**2
+        + X1phis**2 * ddR0 / 4 / R0**2
+    )
+
+    R_2_neg2 = (
+        X1phic * X1phis / 2 / R0
+        + X2Rs
+        + X1Rs * X1phic * dR0 / 2 / R0**2
+        + X1Rc * X1phis * dR0 / 2 / R0**2
+        - X2phis * dR0 / R0
+        - X1phic * X1phis * dR0**2 / R0**3
+        - X1phis * dX1Rc / 2 / R0
+        - X1phic * dX1Rs / 2 / R0
+        + X1phis * dR0 * dX1phic / 2 / R0**2
+        + X1phic * dR0 * dX1phis / 2 / R0**2
+        + X1phic * X1phis * ddR0 / 2 / R0**2
+    )
+
+    R_2_2 = (
+        (X1phic**2 - X1phis**2) / 4 / R0
+        + X2Rc
+        + X1Rc * X1phic * dR0 / 2 / R0**2
+        - X1Rs * X1phis * dR0 / 2 / R0**2
+        - X2phic * dR0 / R0
+        - X1phic**2 * dR0**2 / 2 / R0**3
+        + X1phis**2 * dR0**2 / 2 / R0**3
+        - X1phic * dX1Rc / 2 / R0
+        + X1phis * dX1Rs / 2 / R0
+        + X1phic * dR0 * dX1phic / 2 / R0**2
+        - X1phis * dR0 * dX1phis / 2 / R0**2
+        + X1phic**2 * ddR0 / 4 / R0**2
+        - X1phis**2 * ddR0 / 4 / R0**2
```

Comment:
@dpanici  @f0uriest  R_2_0, R_2_neg2 and R_2_2 can be simplified collecting terms for faster evaluation. Otherwise implementation of second order looks correct. As Rory said, add the Mathematica file or pdf for reference.